Feature/zcs 14115 fix extension dispatcher unregistration issue

RakeshAMore opened the pull request:
**Description**
Extension’s HTTP handlers registered using method https://github.com/Zimbra/zm-mailbox/blob/develop/store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java#L59 can not deregistered using the method https://github.com/Zimbra/zm-mailbox/blob/develop/store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java#L130 since registration of handlers tracking is incorrect, while deregisterd code is not honoring correct key to remove handlers.

**Solution**
Define data structure that can keep tracking of handlers based on the extensions https://github.com/Zimbra/zm-mailbox/blob/develop/store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java#L50 so while removing it will look for handlers with respect to the extensions

synacor-bsinger committed:
ZBUG-382: Reject sieve scripts that are too large.

synacor-bsinger committed:
ZBUG-382: Use globalConfig as the default for zimbraMailSieveScriptMaxSize.

ZBUG-382: Update documentation.

synacor-bsinger committed:
ZBUG-382: Add test for zimbraMailSieveScriptMaxSize global default behavior.

synacor-bsinger committed:
ZBUG-382: Allow reduced-size sieve scripts, even if they're over the size limit.

ZBUG-382: check diff of "require" between old and new sieve scripts

ZBUG-382: modified regex for require statement

ZBUG-382: modified regex for require statement

ZBUG-382: added null check for an argument of Pattern.matcher

preshitaagrawal committed:
ZBUG-382: zimbraMailSieveScript attribute needs to be limited by the number of conditions and rules

ZBUG-382: added SIEVE_SCRIPT_MAX_SIZE_EXCEPTION exception

ZBUG-382: Review comment fixes

shrutig0510 committed:
ZCS-13071 Added NodeId to match uuid in drive document migartion and overloaded saveDocument api

RakeshAMore committed:
ZCS-13901 method for unloading extension

rcyarrapothu committed:
ZCS-13934: updated soap apis to match new license deployment

preshitaagrawal committed:
ZCS-14019: Add LDAP attributes required for 2FA to support email code verification

shubhamCS03 committed:
feature/ZCS-13219:Add an LDAP attribute to hide aliases in GAL

zimsuchitgupta committed:
ZCS-14020: Returing 2FA attribute from Auth response

ZCS-14020: Added condition to retrun masked zimbraPrefPasswordRecoveryAddress if email method is allowed and enabled

k-kato committed:
ZCS-14071: added 2FA objects to ZAuthResult

RakeshAMore committed:
ZCS-14115 Fixing extension dispatcher unregistration issue

dasiyogesh commented on store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java:
```diff
@@ -129,19 +136,17 @@ public void service(HttpServletRequest req, HttpServletResponse resp) throws IOE
-                        it.remove();
-                    }
+            String extName = ext.getName();
+            for (ExtensionHttpHandler extHandler: EXTENSION_HANDLERS.get(extName)) {
```
why for loop, are there more extensions with same name?

RakeshAMore replied:
Loop is for multiple handlers for extension.

dasiyogesh commented on store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java:
```diff
@@ -60,9 +62,14 @@ public static void register(ZimbraExtension ext, ExtensionHttpHandler handler) t
+            }
             sHandlers.put(name, handler);
+            if (!EXTENSION_HANDLERS.containsKey(ext.getName())) {
+                EXTENSION_HANDLERS.put(ext.getName(), new ArrayList<ExtensionHttpHandler>());
```
are there multiple handlers for same extension?

RakeshAMore replied:
Yes.

log2akshat commented on store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java:
```diff
@@ -129,19 +136,17 @@ public void service(HttpServletRequest req, HttpServletResponse resp) throws IOE
+                try {
+                    extHandler.destroy();
+                } catch (Exception e) {
+                    ZimbraLog.extensions.warn("Error in destroy for handler " + extHandler.getClass(), e);
```
`+` operator for string concatenation is an inefficient way to concatenate, it creates new string objects each time, leading to a lot of object creation and garbage collection, which we have seen a lot of problems of memory leaks in Zimbra in the past. And `+` operator can be slower than other methods due to the object creation and garbage collection overhead.

RakeshAMore replied:
Good catch

log2akshat commented on store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java:
```diff
@@ -129,19 +136,17 @@ public void service(HttpServletRequest req, HttpServletResponse resp) throws IOE
+                    extHandler.destroy();
+                } catch (Exception e) {
+                    ZimbraLog.extensions.warn("Error in destroy for handler " + extHandler.getClass(), e);
+                } finally {
```
Use `try-with-resources `here, as it automatically handles cleanup, improves code readability, enhances exception handling, reduces resource leak risks, and simplifies null checking compared to traditional `finally` blocks.

RakeshAMore replied:
That's good in other cases. but here it seems `try-catch-finally` is the best choice. since we are not opening try-on `sHandlers`

log2akshat commented on store/src/java/com/zimbra/cs/extension/ExtensionDispatcherServlet.java:
```diff
@@ -48,6 +47,9 @@
 public class ExtensionDispatcherServlet extends ZimbraServlet {
     private static Map sHandlers = Collections.synchronizedMap(new HashMap());
+
+    private static Map<String, ArrayList<ExtensionHttpHandler>> EXTENSION_HANDLERS = Collections.synchronizedMap(new HashMap());
```
For improved performance in high-concurrency scenarios, it is generally recommended to use `ConcurrentHashMap` over a `synchronizedMap`, as it employs fine-grained locking and provides efficient concurrent access to map data.

RakeshAMore replied:
hmm. that's true, since we are not EXTENSION_HANDLERS as lock here we can go ahead with the ConcurrentHashMap here.

zimsuchitgupta committed:
ZCS-14018: Added request and response file for SendTFACode

chitranshu198 committed:
ZBUG-3132 : Changes to fix missing external email warning issue for all external domains

preshitaagrawal committed:
ZCS-14058 Update EnableTwoFactorAuthRequest and DisableTwoFactorAuthRequest to address email method

preshitaagrawal committed:
ZCS-14022 Refactor SetRecoveryAccountRequest to handle zimbraPrefPasswordRecoveryAddress for 2FA email

zimsuchitgupta committed:
ZCS-14191: Refactored classes and added auth token in request body

Commit:
Zbug-3621- added pkcs7-mime and pkcs7-signature mime types in custom-mimetypes.xml and mapped respective patterns.

sumitkumar1110 committed:
ZBUG-3565 - Added logs for audit.log and mailbox.log for non-existing users

rcyarrapothu committed:
ZCS-14106 : adding ldap attributes for chat-video functionality

shubhamCS03 committed:
ZCS-14416:Date filter payload modification due to timezone issue

zimsuchitgupta committed:
ZCS-14379: Sanitized attribute zimbraTwoFactorCodeForEmail value to VALUE-BLOCKED

zimsuchitgupta committed:
ZCS-14369: Defined action and status as Element instead of attribute

zimsuchitgupta committed:
ZCS-14364: Provided default template for zimbraTwoFactorCodeEmailBody|Subject

zimsuchitgupta committed:
ZCS-14368: Configured email code length by adding a ldap attribute

shrutig0510 committed:
ZCS-14422 Added inviteId on invite instance

k-kato committed:
ZCS-14405: added new LDAP attribute zimbraServerVersionChangeNotificationDisabled

shubhamCS03 committed:
ZCS-14294:Social filter is not working with latest URL's of social platforms

oscarbrookssynacor committed:
ZCS-14424 : Limit attribute restriction with LDAP Counter: LdapCounter

k-kato committed:
ZCS-14375: update translation for 2FA

Commit:
ZCS-14474: ZBUG-2956 - Send email button does not work in modern UI with Undo send enable - delegate sender Send as

Added test case for scheduleTask Method.

Updated according to review comments

Changed the comments

Updated the comments

zimsuchitgupta committed:
ZCS-14435: Added request and response classes for DisableTwoFactorAuth for admin namespace

k-kato committed:
ZCS-14429: updated optionalIn of zimbraPrefPrimaryTwoFactorAuthMethod

Commit:
ZCS-14564: PDF attachment icon shows as plain text for the PDFs having display issue.

zimsuchitgupta committed:
ZCS-14458: Added zimbraTwoFactorAuthMethodAllowed info in exception message

zimsuchitgupta committed:
ZCS-14476: Renamed and added an attribute to define html email body

k-kato committed:
ZCS-14531: fixed id and since of some attributes for 2FA

Commit:
ZCS-14643: Adding Check for Null Pointer Exception.

shubhamCS03 committed:
ZCS-14611: User is able to delete account when license is in grace period

Commit:
ZCS-14653: Add null Pointer check for delegatorAccount variable.

Updated according to review comments

Commit:
ZBUG-2859 Bugzilla 80485 admin login honors zimbraAuthFallbackToLocal

Commit:
Revert "ZCS-14653: Add null Pointer check for delegatorAccount variable."

This reverts commit 05c3fa183897b6280e72d6140e025e8922003d2d.

Commit:
Revert "ZCS-14643: Adding Check for Null Pointer Exception."

This reverts commit ec6b630256b77935023e6e07708fb57618539259.

Commit:
Revert "ZCS-14474: ZBUG-2956 - Send email button does not work in modern UI with Undo send enable - delegate sender Send as"

This reverts commit 16110f65d03e9b50665f10d32ae2cdc092c66b20.

swathy-shaji committed:
ZCS-14697: Jenkins build is getting failed after merging the fix for user account deletion during grace period

vishnukantjangid committed:
ZBUG-3836: CLI support for compression on S3 storage external volumes

rendurama committed:
ZCS-14704 Manual Cache refresh command for license cache

zimsuchitgupta committed:
ZCS-14647: Added SendTwoFactorCodeRequest/Response APIs for admin namespace space

rcyarrapothu committed:
ZCS-14707 : intial version - offline activation through LAN Daemon

zimsuchitgupta committed:
ZCS-14701: Refactored code and returing 2FA attributes

swathy-shaji committed:
ZCS-14680: Usage Reporting

shubhamCS03 committed:
ZCS-13219-1: Updaing the GAL attribute Id

miteshsavani810 committed:
PREAPPS-7711: Add new LDAP attributes that require to disable Classic UI access
- Added new zimbraClassicWebClientDisabled attribute at domain and config level.

shrutig0510 committed:
ZCS-14787 Set zimbraMobileForceSamsungProtocol25 default to TRUE

swathy-shaji committed:
ZCS-14766: [Usage Reporting] consider the counter for COS and Domain level

zimsuchitgupta committed:
ZCS-14592: Setting account specific timezone

zimsuchitgupta committed:
ZCS-14407: Added callback for zimbraTwoFactorScratchCodeLength and zimbraTwoFactorCodeLength

ZCS-14407: Addressed review comments

Commit:
ZCS-14782: IP based restriction - Add LDAP attributes

rcyarrapothu committed:
ZCS-14767 : added clear-certificate api

Commit:
ZCS-14781: Add ldap attributes to reset TFA setting

Updated according to review comment.

Updated according to review comments.

rcyarrapothu committed:
ZCS-14893 : added a new attribute instead of replacing older one

RakeshAMore committed:
ZCS-14904 creating new LDAP attribute for licensing

dasiyogesh committed:
ZBUG-3987: Set AllowConcurrentSocketFactoryUse to true for LDAP connections

k-kato committed:
ZCS-14859: added exceptions related to 2FA

shrutig0510 committed:
ZBUG-3863 Gracefully handling OOM for reindex account (#1592)

* ZBUG-3863 Gracefully handling OOM for reindex account

* ZBUG-3863 replacing empty variable

miteshsavani810 committed:
ZCS-14933: Change in clientInfo API to get value of zimbraClassicWebClientDisabled attribute.

rcyarrapothu committed:
ZCS-14969 : separate response string for act certs

swathy-shaji committed:
ZCS-14877: Adding Feature Grace Period Validation while creating and modify an Account or COS

zimsuchitgupta committed:
ZCS-14964: Proxying request to server where it exist

rcyarrapothu committed:
ZCS-14979 : chat-video license attributes update

preshitaagrawal committed:
ZCS-15010: Add LDAP attribute to hold email address for Licensing notification to be sent to

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Zimbra/zm-mailbox?pullRequest=1525) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Zimbra/zm-mailbox?pullRequest=1525) before we can accept your contribution.<br/>**12** out of **17** committers have signed the CLA.<br/><br/>:white_check_mark: rcyarrapothu<br/>:white_check_mark: RakeshAMore<br/>:white_check_mark: k-kato<br/>:white_check_mark: shrutig0510<br/>:white_check_mark: swathy-shaji<br/>:white_check_mark: rendurama<br/>:white_check_mark: shubhamCS03<br/>:white_check_mark: zimsuchitgupta<br/>:white_check_mark: miteshsavani810<br/>:white_check_mark: barrydegraaff<br/>:white_check_mark: dasiyogesh<br/>:white_check_mark: vishnukantjangid<br/>:x: preshitaagrawal<br/>:x: sumitkumar1110<br/>:x: chitranshu198<br/>:x: oscarbrookssynacor<br/>:x: synacor-bsinger<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Zimbra/zm-mailbox?pullRequest=1525) it.</sub>
